test/e2e_node/node_shutdown_linux_test.go

```diff
@@ -207,9 +208,187 @@ var _ = SIGDescribe("GracefulNodeShutdown [Serial] [NodeAlphaFeature:GracefulNod
+				_, err := f.ClientSet.SchedulingV1().PriorityClasses().Create(context.Background(), customClass, metav1.CreateOptions{})
+				framework.ExpectNoError(err)
+			}
+		})
+
+		ginkgo.AfterEach(func() {
+			ginkgo.By("Emitting Shutdown false signal; cancelling the shutdown")
+			err := emitSignalPrepareForShutdown(false)
+			framework.ExpectNoError(err)
+		})
+
+		ginkgo.It("should be able to gracefully shutdown pods with various grace periods", func() {
+			nodeName := getNodeName(f)
+			nodeSelector := fields.Set{
+				"spec.nodeName": nodeName,
+			}.AsSelector().String()
+
+			// Define test pods
+			pods := []*v1.Pod{
+				getGracePeriodOverrideTestPod("period-120", nodeName, 120, ""),
+				getGracePeriodOverrideTestPod("period-5", nodeName, 5, ""),
+				getGracePeriodOverrideTestPod("period-c-120", nodeName, 120, customClassC.Name),
+				getGracePeriodOverrideTestPod("period-c-5", nodeName, 5, customClassC.Name),
+				getGracePeriodOverrideTestPod("period-b-120", nodeName, 120, customClassB.Name),
+				getGracePeriodOverrideTestPod("period-b-5", nodeName, 5, customClassB.Name),
+				getGracePeriodOverrideTestPod("period-a-120", nodeName, 120, customClassA.Name),
+				getGracePeriodOverrideTestPod("period-a-5", nodeName, 5, customClassA.Name),
+				getGracePeriodOverrideTestPod("period-critical-120", nodeName, 120, scheduling.SystemNodeCritical),
+				getGracePeriodOverrideTestPod("period-critical-5", nodeName, 5, scheduling.SystemNodeCritical),
+			}
+
+			// Expected down steps
+			downSteps := [][]string{
+				{
+					"period-5", "period-120",
+				},
+				{
+					"period-5", "period-120",
+					"period-c-5", "period-c-120",
+				},
+				{
+					"period-5", "period-120",
+					"period-c-5", "period-c-120",
+					"period-b-5", "period-b-120",
+				},
+				{
+					"period-5", "period-120",
+					"period-c-5", "period-c-120",
+					"period-b-5", "period-b-120",
+					"period-a-5", "period-a-120",
+				},
+				{
+					"period-5", "period-120",
+					"period-c-5", "period-c-120",
+					"period-b-5", "period-b-120",
+					"period-a-5", "period-a-120",
+					"period-critical-5", "period-critical-120",
+				},
+			}
+
+			ginkgo.By("Creating batch pods")
+			f.PodClient().CreateBatch(pods)
+
+			list, err := f.PodClient().List(context.TODO(), metav1.ListOptions{
+				FieldSelector: nodeSelector,
+			})
+			framework.ExpectNoError(err)
+			framework.ExpectEqual(len(list.Items), len(pods), "the number of pods is not as expected")
+
+			ginkgo.By("Verifying batch pods are running")
+			for _, pod := range list.Items {
+				if podReady, err := testutils.PodRunningReady(&pod); err != nil || !podReady {
+					framework.Failf("Failed to start batch pod: %v", pod.Name)
+				}
+			}
+
+			ginkgo.By("Emitting shutdown signal")
+			err = emitSignalPrepareForShutdown(true)
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Verifying that pods are shutdown")
+
+			for _, step := range downSteps {
+				gomega.Eventually(func() error {
```

Comment:
ping on above question that also got resolved accidentally: 

I'm curious your thinking here ---

as I understand we're trying to e2e test that at a certain moment only some subset of pods are shutdown. However the shutdown and e2e checks here are not synchronized, so it's very possible that the shutdown will move faster than validation. As a result, it's possible that more pods will be shutdown than expected at certain "step". Is that handled somehow or is that a non issue?

Reply:
> However the shutdown and e2e checks here are not synchronized, so it's very possible that the shutdown will move faster than validation. As a result, it's possible that more pods will be shutdown than expected at certain "step". Is that handled somehow or is that a non issue?

https://github.com/kubernetes/kubernetes/blob/a17703894978835bd7cf00ceadb4a7587205c61c/test/e2e_node/node_shutdown_linux_test.go#L350
In L350, Every time it is judged whether the number of obtained Pods meets expectations, this should not happen.

Reply:
ack makes sense. However, it seems like the shutdown is initiated at `emitSignalPrepareForShutdown(true)`  on line 337, and then the loop logic is: 

```
for _, step := range downSteps { 
    gomega.Eventually(func() 
       ....
```

so to me, it seems like potentially depending on how fast kubelet does the shutdown and how long the grace periods are, the kubelet has possibility of moving faster then the checks on the test. 

For example it could lead to situation where more pods are shutdown than expected at some moment in time due to `downStep` not being synchronized with kubelet pod shutdown. 

Not sure if it's real issue though, we can always increase the grace period and see how it works in the testgrid...